### DISCUSSION TITLE
eslint configuration: ecmaVersion 5 or 6 or else?

### DISCUSSION THREAD

**Discussion by HiranChaudhuri** [eslint/eslint#17896] [General]
Based on issue https://github.com/eslint/eslint/issues/17882, which is unfortunately closed already:

Thank you for the good explanation. I switched the linter to JavaScript 6 - and as you indicated the parsing error was gone.

I am coming back though since I found out the project is running on SpiderMonkey 1.8.5. This version actually supports ECMASCRIPT 5 with the exception that the destructuring syntax is already supported.

So would it be possible to enable eslint on ECMASCRIPT 5 plus configure an additional/custom parsing rule, or would it be more suitable to just switch to ECMASCRIPT 6? After all linting towards 6 could bring some deviations not understood by the runtime of the project.

**Comment by mdjermanovic** [comment]
No, it's not possible to configure Espree (ESLint's default parser) to allow ES5 + destructing only.

What you could try is to switch parsing to ES6 and enable rules from [eslint-plugin-es-x](https://github.com/eslint-community/eslint-plugin-es-x) that disallow ES6 syntax other than destructuring.

**Reply by HiranChaudhuri** [reply]
Seems like eslint-plugin-es-x is perfect for my case. I checked the [documentation usage example](https://eslint-community.github.io/eslint-plugin-es-x/#%F0%9F%93%96-usage) and like it.
There is then only one question: How would I find the list of rules that I need to disallow?

Would I have to configure ES2015, then list all rules from the ES2015 group but es-x/no-destructuring ?

**Reply by HiranChaudhuri** [reply]
Ok, while writing the configuration I understood what needs to happen. Just in case others need similar stuff, here is the configuration I created. Sorry, the excerpt of .eslintrc.json is a bit lengthy...

```
    "plugins": ["es-x"],
    "parserOptions": {
        "ecmaVersion": 2015
    },
    "rules": {
        "no-mixed-spaces-and-tabs": "off",
        "no-unused-vars": ["error", { "argsIgnorePattern": "^_"}],
        "es-x/no-array-from": "error",
        "es-x/no-array-of": "error",
        "es-x/no-array-prototype-copywithin": "error",
        "es-x/no-array-prototype-entries": "error",
        "es-x/no-array-prototype-fill": "error",
        "es-x/no-array-prototype-find": "error",
        "es-x/no-array-prototype-findindex": "error",
        "es-x/no-array-prototype-keys": "error",
        "es-x/no-array-prototype-values": "error",
        "es-x/no-arrow-functions": "error",
        "es-x/no-binary-numeric-literals": "error",
        "es-x/no-block-scoped-functions": "error",
        "es-x/no-block-scoped-variables": "error",
        "es-x/no-classes": "error",
        "es-x/no-computed-properties": "error",
        "es-x/no-default-parameters": "error",
        "es-x/no-destructuring": "off",
        "es-x/no-for-of-loops": "error",
        "es-x/no-generators": "error",
        "es-x/no-map": "error",
        "es-x/no-math-acosh": "error",
        "es-x/no-math-asinh": "error",
        "es-x/no-math-atanh": "error",
        "es-x/no-math-cbrt": "error",
        "es-x/no-math-clz32": "error",
        "es-x/no-math-cosh": "error",
        "es-x/no-math-expm1": "error",
        "es-x/no-math-fround": "error",
        "es-x/no-math-hypot": "error",
        "es-x/no-math-imul": "error",
        "es-x/no-math-log10": "error",
        "es-x/no-math-log1p": "error",
        "es-x/no-math-log2": "error",
        "es-x/no-math-sign": "error",
        "es-x/no-math-sinh": "error",
        "es-x/no-math-tanh": "error",
        "es-x/no-math-trunc": "error",
        "es-x/no-modules": "error",
        "es-x/no-new-target": "error",
        "es-x/no-number-epsilon": "error",
        "es-x/no-number-isfinite": "error",
        "es-x/no-number-isinteger": "error",
        "es-x/no-number-isnan": "error",
        "es-x/no-number-issafeinteger": "error",
        "es-x/no-number-maxsafeinteger": "error",
        "es-x/no-number-minsafeinteger": "error",
        "es-x/no-number-parsefloat": "error",
        "es-x/no-number-parseint": "error",
        "es-x/no-object-assign": "error",
        "es-x/no-object-getownpropertysymbols": "error",
        "es-x/no-object-is": "error",
        "es-x/no-object-setprototypeof": "error",
        "es-x/no-object-super-properties": "error",
        "es-x/no-octal-numeric-literals": "error",
        "es-x/no-promise": "error",
        "es-x/no-property-shorthands": "error",
        "es-x/no-proxy": "error",
        "es-x/no-reflect": "error",
        "es-x/no-regexp-prototype-flags": "error",
        "es-x/no-regexp-u-flag": "error",
        "es-x/no-regexp-y-flag": "error",
        "es-x/no-rest-parameters": "error",
        "es-x/no-set": "error",
        "es-x/no-spread-elements": "error",
        "es-x/no-string-fromcodepoint": "error",
        "es-x/no-string-prototype-codepointat": "error",
        "es-x/no-string-prototype-endswith": "error",
        "es-x/no-string-prototype-includes": "error",
        "es-x/no-string-prototype-normalize": "error",
        "es-x/no-string-prototype-repeat": "error",
        "es-x/no-string-prototype-startswith": "error",
        "es-x/no-string-raw": "error",
        "es-x/no-subclassing-builtins": "error",
        "es-x/no-symbol": "error",
        "es-x/no-template-literals": "error",
        "es-x/no-typed-arrays": "error",
        "es-x/no-unicode-codepoint-escapes": "error",
        "es-x/no-weak-map": "error",
        "es-x/no-weak-set": "error"
    },

```

**Reply by mdjermanovic** [reply]
Another way that might work is to [use](https://eslint.org/docs/latest/use/configure/configuration-files#using-a-configuration-from-a-plugin) one of the [configs this plugin provides](https://eslint-community.github.io/eslint-plugin-es-x/#presets) (I think `restrict-to-es5` or `no-new-in-es2015`; probably the latter since you'd likely only need rules that disallow ES2015 features, but I've never used this plugin so please double check) and just disable the rule that disallows destructuring (`"es-x/no-destructuring": "off"`).

**Reply by HiranChaudhuri** [reply]
That would make the whole config a lot more readable. Let me see when I get to check that out.
